README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# vinference

<!-- badges: start -->

[![Lifecycle:
maturing](https://img.shields.io/badge/lifecycle-maturing-blue.svg)](https://www.tidyverse.org/lifecycle/#maturing)
[![Last-changedate](https://img.shields.io/badge/last%20change-2020--10--10-yellowgreen.svg)](/commits/master)
[![Travis build
status](https://travis-ci.com/heike/vinference.svg?branch=master)](https://travis-ci.com/heike/vinference)
<!-- badges: end -->

The goal of vinference is to assist with the design and analysis of
visual inference experiments.

For more information on visual inference, check out the `nullabor` R
package or [Wickham et
al. (2010)](https://doi.org/10.1109/TVCG.2010.161).

## Installation

`vinference` is not yet on CRAN; you can install the development version
from [GitHub](https://github.com/) with:

``` r
# install.packages("devtools")
devtools::install_github("heike/vinference")
```

## Example

``` r
library(vinference)
```

In visual inference, we can run experiments under 3 scenarios:

1.  Scenario 1: in each of K evaluations a different data set and a
    different set of (m-1) null plots is shown.
2.  Scenario 2: in each of K evaluations the same data set but a
    different set of (m-1) null plots is shown.
3.  Scenario 3: the same lineup, i.e. same data and same set of null
    plots, is shown to K different observers.

To calculate a p-value for a set of 20-panel single-target lineups
generated under Scenario 1 where the data plot was selected 10 times out
of 30 observations,

``` r
pVis(10, K = 30, m = 20, scenario = 1, lower.tail = F)
#> [1] 1.161521e-06
```

Under Scenario 3, however, we would use the same 20-panel single-target
lineup and have 30 different participants evaluate it. Because we are
using the same panels for each participant evaluation, there is
dependence between the evaluations of different participants and we must
account for this using a parameter \(\alpha\) which provides information
about the level of visual interest of the null panels. See XXX Stat
paper citation for more information.

Suppose, for now, we know \(\alpha \approx 0.1\). Then our p-value for
the lineup would be

``` r
pVis(10, K = 30, m = 20, scenario = 3, alpha = 0.1, lower.tail = F)
#> [1] 0.2232121
```

Probability (`dVis`) and quantile (`qVis`) functions are also available.

As discussed in XXX Cite stat paper XXX, we need to estimate the
hyperparameter \(\alpha\), which describes the number of visually
interesting null plots generated by the null model.

This approach is appropriate when you have several single or dual target
lineup evaluations with similar (non-zero) number of null panel
selections. Eliminate target panel selections and work with the
remaining \(K\) total null panel selections and a \(m_0\) equal to the
number of null panels. Determine how many selections \(c\) are required
for a panel to be considered interesting (by default, we recommend
\(K/m_0\)). Using your values for \(c\), \(K\) and \(m_0\), the
`estimate_alpha_visual` function will create a plot with the expected
number of null panels selected at least \(c\) times; simulated
experiments will be plotted around the line to provide some measure of
variability.

``` r
estimate_alpha_visual(m0 = 19, K = 30, c = 1)
```

<img src="man/figures/README-unnamed-chunk-3-1.png" width="100%" /> If
you have experimental data available, there is an additional function
which may assist you with visual estimation: `observed_band`. The
observed band function leverages an average number of panels or a range
of panels to add additional cues to the plot. Suppose that we have
several lineups which have \(K\approx30\) evaluations; between 4 and 7
panels were selected in each lineup, with an average of 4.75 panels
selected.

``` r
estimate_alpha_visual(m0 = 19, K = 30, c = 1) + observed_band(4.75, c(4, 7))
#> Warning: Transformation introduced infinite values in continuous x-axis

#> Warning: Transformation introduced infinite values in continuous x-axis
```

<img src="man/figures/README-unnamed-chunk-4-1.png" width="100%" />
Using this cue, values of \(\alpha\) between 0.075 and 0.12 might be
most appropriate, with the best numerical estimate around
\(\alpha=0.1\).

We can then examine whether our p-value of 9.67049710^{-4} corresponding
to \(\alpha=0.1\) results in a similar conclusion for all plausible
\(\alpha\) values in our observed range. At \(\alpha=0.0175\), the
corresponding p-value is 0.0015002. At \(\alpha=0.012\), the
corresponding p-value is 6.642673610^{-4}. In both cases, our lineup
plot is significant, and we come to the same conclusion: the target plot
is a significant visual statistic.


NEWS
============================================================
Changes across different versions
============================================================
------------------------------------------------------------
Version 1.0.0
------------------------------------------------------------


R/multi.r
#' @include pvalues.r
dvismulti1 <- function(K, k, m = 20, N = 5000) {
  stopifnot((K == length(k)) | (length(k) == 1))
  k <- rep(k, length = K)
  res <- replicate(N, {
    # everybody is shown a different lineup
    #   require(plyr)
    success <- purrr::map_dbl(1:K, function(i) {
      lp <- stats::runif(m)
      sum(sample(1:m, size = k[i], replace = FALSE, prob = lp) == 1)
    })
    # success <- plyr::ldply(1:K, )
    sum(success)
  })
  res <- factor(res, levels = 0:K)
  table(res) / N
}

dvismulti2 <- function(K, k, m = 20, N = 5000) {
  stopifnot((K == length(k)) | (length(k) == 1))
  k <- rep(k, length = K)
  res <- replicate(N, {
    # same data, different nulls
    #   require(plyr)
    first <- stats::runif(1)
    success <- purrr::map_dbl(1:K, function(i) {
      lp <- c(first, stats::runif(m - 1))
      sum(sample(1:m, size = k[i], replace = FALSE, prob = lp) == 1)
    })
    # success <- plyr::ldply(1:K, function(i) {
    #   lp <- c(first, runif(m-1))
    #   sum(sample(1:m, size=k[i], replace=FALSE, prob=lp)==1)
    # })
    sum(success)
  })
  res <- factor(res, levels = 0:K)
  table(res) / N
}
dvismulti3 <- function(K, k, m = 20, N = 5000) {
  stopifnot((K == length(k)) | (length(k) == 1))
  k <- rep(k, length = K)
  res <- replicate(N, {
    # everybody is shown the same lineup
    lp <- stats::runif(m)
    #   require(plyr)
    success <- purrr::map_dbl(1:K, function(i) {
      sum(sample(1:m, size = k[i], replace = FALSE, prob = lp) == 1)
    })
    # success <- plyr::ldply(1:K, function(i) {
    #   sum(sample(1:m, size=k[i], replace=FALSE, prob=lp)==1)
    # })
    sum(success)
  })
  res <- factor(res, levels = 0:K)
  table(res) / N
}




#' Simulation based density and distribution functions of visual inference under a multiple choice lineup
#'
#' Simulation based p value to observe x or more picks of the data plot in K evaluations of a multiple choice lineup test  under the assumption that the data plot is consistent with the null hypothesis.
#' We distinguish between three different scenarios:
#' \itemize{
#' \item Scenario I: in each of K evaluations a different data set and a different set of (m-1) null plots is shown.
#' \item Scenario II: in each of K evaluations the same data set but a different set of (m-1) null plots is shown.
#' \item Scenario III: the same lineup, i.e. same data and same set of null plots, is shown to K different observers.
#' }
#' @name vmulti
#' @aliases dmulti pmulti qmulti
#' @param x (vector) of the number of observed data picks
#' @param q (vector) of the number of observed data picks
#' @param K integer value, specifying the number of participants/evaluations
#' @param k (vector) of integer values of the number of observed picks in a multiple choice lineup evaluation
#' @param m lineup size, defaults to m=20
#' @param type character, one of "scenario1", "scenario2", or "scenario3"
#' @param N integer value, number of simulations.
#' @return list consisting of two named items:
#' \itemize{
#' \item dmulti: density estimates and vector k
#' \item pmulti: distribution estimates and vector k
#' \item qmulti: quantile estimates and vector k
#' }
#' @examples
#' k=rpois(5,lambda=1)+1
#' m=20
#' dmulti(0:5,K=5,k=k, type="scenario1", m)
#' ## compare to Poisson Binomial:
#' library(poibin)
#' dpoibin(0:5, pp=k/m)
#'
#' qmulti(c(0.95, 0.99), K=5, k=2)
#'
#' \dontrun{
#' (k <- rpois(5, lambda=1.5)+1)
#' reps1 <- plyr::ldply(1:10, function(x) dmulti(0:5, 5, k, type="scenario1")$density)
#' reps2 <- plyr::ldply(1:10, function(x) dmulti(0:5, 5, k, type="scenario2")$density)
#' reps3 <- plyr::ldply(1:10, function(x) dmulti(0:5, 5, k, type="scenario3")$density)
#' reps1$type <- "I"
#' reps2$type <- "II"
#' reps3$type <- "III"
#' reps <- rbind(reps1, reps2, reps3)
#' library(reshape2)
#' library(RColorBrewer)
#' cols <- brewer.pal(8, "Paired")
#'
#' mr <- melt(reps, measure.vars=1:6)
#' mr$variable <- as.numeric(gsub("V", "", mr$variable))
#' mr$type <- factor(mr$type)
#' mr$offset <- with(mr, 0.1*c(0,-1,1)[as.numeric(type)])
#' qplot(variable+offset, value, data=mr, alpha=I(0.75), colour=type) +
#'   geom_point(aes(x=1:6, y=dpoibin(0:5, pp=k*1/m)), pch=1, size=4,
#'     inherit.aes=FALSE)+
#'   xlab("x")+ylab("P(X=x)")+ggtitle(paste(k, collapse=",")) +
#'   theme_bw() + scale_colour_manual(values=cols[-c(1,3,5,6,7)])
#'  }


NULL

#' @rdname vmulti
#' @aliases dmulti pmulti qmulti
#' @export
dmulti <- function(x, K, k, m = 20, type = "scenario3", N = 5000) {
  density <- switch(type, scenario1 = dvismulti1(K, k, m, N),
    scenario2 = dvismulti2(K, k, m, N),
    scenario3 = dvismulti3(K, k, m, N)
  )
  res <- list(density = as.vector(density)[x + 1], k = k)

  res
}

#' @rdname vmulti
#' @aliases dmulti pmulti qmulti
#' @export
pmulti <- function(x, K, k, m = 20, type = "scenario3", N = 5000) {
  density <- switch(type, scenario1 = dvismulti1(K, k, m, N),
    scenario2 = dvismulti2(K, k, m, N),
    scenario3 = dvismulti3(K, k, m, N)
  )
  res <- list(distribution = as.vector(cumsum(density))[x + 1], k = k)

  res
}

#' @rdname vmulti
#' @aliases dmulti pmulti qmulti
#' @export
qmulti <- function(q, K, k, m = 20, type = "scenario3", N = 5000) {
  density <- switch(type, scenario1 = dvismulti1(K, k, m, N),
    scenario2 = dvismulti2(K, k, m, N),
    scenario3 = dvismulti3(K, k, m, N)
  )

  res <- list(quantile = cumsum(as.vector(density)), k = k)
  # res$quantile <- plyr::laply(q, function(qq) min(which(res$quantile >= qq)-1))
  res$quantile <- purrr::map_dbl(q, function(qq) min(which(res$quantile >= qq) - 1))
  names(res$quantile) <- q

  res
}


R/alpha_ml.R
#' Estimate alpha of a symmetric Dirichlet distribution
#' 
#' The estimation of the rate parameter alpha is done with Maximum Likelihood Estimation.
#' @param p (matrix) of probabilities. Number of columns is interpreted as dimension of the Dirichlet distribution
#' Number of rows is repetitions. 
#' @param weight vector
#' @param eps error allowed in finding root
#' @return rate parameter alpha of symmetric (flat) Dirichlet distribution
#' @importFrom stats rgamma uniroot
#' @export
#' @examples
#' x <- matrix(rgamma(n=2000, shape=.1), ncol=20)
#' plot(density(x[,1]))
#' x <- x/rowSums(x)
#' alpha_ml(x)
#' shapes <- seq(0.01, 0.75, by = 0.01)
#' ests <- sapply(shapes, function(s) {
#'   x <- matrix(rgamma(n=2000, shape=s), ncol=20)
#'   alpha_ml(x, eps=10^-20)
#' })
#' plot(shapes,ests)
#' abline(a=0,b=1)
alpha_ml <- function(p, weight=NULL, eps=10^(-7)) {
  # matrix p
  if (is.null(dim(p))) {
    m <- length(p)
    n <- 1
    p <- matrix(p, nrow=1)
  } else {
    n <- dim(p)[1]
    m <- dim(p)[2]
  }
  if (is.null(weight)) weight <- rep(1, n)
  else n <- sum(weight)
  
  weight <- weight/sum(weight)
  ps <- p
  ps[p < eps] <- eps # make sure we don't take the log of 0
  ps <- ps/rowSums(ps)
  logp <- sum(weight*rowSums(log(ps)))/m
  ml <- function(alpha) {
    digamma(alpha) - digamma(alpha*m) - logp
  }
  # find alpha such that digamma(alpha) - digamma(alpha*m) = logp
  alpha <- uniroot(f=ml, interval=c(eps,10))
  
  
  alpha$root
}


R/VK33-resolve.R
#' @include VK33-recursion.R
p3r <- function(K, x) {
  #  require(plyr)
  # as.vector(choose(K,x)*unlist(plyr::llply(x, function(y) h3r(x=y, K=K))))
  as.vector(choose(K, x) * purrr::map_dbl(x, function(y) h3r(x = y, K = K)))
}

h3r <- function(K, x) {
  if (K < 0) return(NA)
  if (K == 0) return(1)
  if (K == 1) {
    if (x == 0) return(2 / 3)
    if (x == 1) {
      return(1 / 3)
    } else {
      return(NA)
    }
  }
  if (x == 0) {
    return((1 - g3(K, 0)) / (K - 1))
  }
  i <- 2:x
  pp <- prod(i / (K - 1 + i - x))
  if (2 > x) pp <- 1

  gi <- rep(NA, x)
  for (i in 1:x) {
    j <- (i + 1):x
    ppj <- prod(j / (K - 1 - x + j))
    if (i + 1 > x) ppj <- 1
    gi[i] <- ppj * g3(K + i - x, i) / (K - x + i - 1)
  }

  if (x < K) {
    h3Kx <- pp / (K - x) * h3r(K - x, 0) - sum(gi)
  } else {
    h3Kx <- pp * h3r(1, 1) - sum(gi[-1])
  }
  return(h3Kx)
}



g3 <- function(K, x) {
  if (K < 2) return(NA)
  if (K == 2) {
    if (x == 0) return(log(27 / 16))
    if (x == 1) return(1 - log(27 / 16))
    if (x == 2) {
      return(log(27 / 16))
    } else {
      return(NA)
    }
  }
  if (x == K) {
    if (K == 3) return(f3(3, 3))
    return((27 / 3^K - 16 / 2^K + 1) / ((K - 2) * (K - 3)))
  }
  return((f3(K, x) + (K - x) * g3(K - 1, x)) / (K - 2))
}



f3 <- function(K, x) {
  if (K < 2) return(NA)
  if (K == 2) {
    if (x == 0) return(-2)
    if (x == 1) return(-1)
    if (x == 2) {
      return(0)
    } else {
      return(NA)
    }
  }
  if (K == 3) {
    if (x == 0) return(-log(4 / 3) - 1)
    if (x == 1) return(log(4 / 3) - 1)
    if (x == 2) return(-log(4 / 3))
    if (x == 3) {
      return(log(4 / 3))
    } else {
      return(NA)
    }
  }
  if (x == K) {
    return((1 - 2 / 2^(K - 3) + 1 / 3^(K - 3)) / (K - 3))
  }
  return((-2^(4 - K) + 2^(K - x) / 3^(K - 3) + (K - x) * f3(K - 1, x)) / (K - 3))
}


R/pvalues-deprecated.r
# data(sysdata, envir=environment())

lineup <- function(m, dataprob = NULL, nulls = NULL) {
  # assume first element is data
  probs <- stats::runif(m)
  n.targets <- length(dataprob)

  if (!is.null(dataprob)) probs[1:n.targets] <- dataprob
  if (!is.null(nulls)) probs[(1 + n.targets):m] <- nulls
  sample(m, size = 1, prob = 1 - probs)
}

pickData <- function(m, xp = 1, dataprob = NULL, nulls = NULL) {
  probs <- stats::runif(m)
  n.targets <- length(dataprob)
  if (!is.null(dataprob)) {
    probs[1:n.targets] <- dataprob
  }
  if (!is.null(nulls)) probs[(n.targets + 1):m] <- nulls
  #  sample(m, size=1, prob=1-probs)
  #  stats::rbinom(1, size=1, prob=f(probs))
  ps <- (1 - probs)^xp
  if (all(ps == 0)) ps <- rep(1, length(probs))
  stats::rbinom(1, size = 1, prob = sum(ps[1:n.targets]) / sum(ps))
}

# too deterministic
# function(ps) ps[1] < min(ps[-1]))
# just wrong:
# function(ps) stats::rbinom(1, size=1, prob=(1-ps[1])/(2-ps[1] - min(ps[-1])))

scenario1 <- function(N, K, m = 20, xp = 1, target = 1) {
  # new lineup in each evaluation: new data, new sample of nulls
  table(replicate(N, {
    individual <- sum(replicate(K, pickData(m, dataprob = NULL, xp = xp)) %in% target)
    individual
  })) / N
}


scenario2 <- function(N, K, m = 20, xp = 1, target = 1) {
  # each data evaluated K times, always with different nulls
  table(replicate(N, {
    n.targets <- length(target)
    dataprob <- stats::runif(n.targets)
    individual <- sum(replicate(K, pickData(m, dataprob = dataprob, xp = xp) %in% target))
    individual
  })) / N
}


scenario3 <- function(N, K, m = 20, xp = 1, target = 1) {
  # each data evaluated K times, with the same nulls
  table(replicate(
    N / 100,
    replicate(100, {
      n.targets <- length(target)
      dataprob <- stats::runif(n.targets)
      nulls <- stats::runif(m - n.targets)

      individual <- sum(replicate(K, pickData(m, dataprob = dataprob, nulls = nulls, xp = xp)) %in% target)
      individual
    })
  )) / N
}

scenario4 <- function(N, K, m = 20, xp = 1, target = 1) {
  # K is vector: length(K) lineups are shown K[i] times to observers
  # all length(K) lineups show the same data, but have different nulls

  res <- replicate(N, {
    n.targets <- length(target)
    dataprob <- stats::runif(n.targets)

    individual <- rep(NA, length(K))
    for (i in 1:length(K)) {
      nulls <- stats::runif(m - n.targets)
      individual[i] <- sum(replicate(K[i], pickData(m, dataprob = dataprob, nulls = nulls, xp = xp)) %in% target)
    }
    sum(individual)
  })
  table(res) / N
}

#' Bootstrap based p values for visual inference
#'
#' \lifecycle{deprecated}
#' Simulation based p value to observe x or more picks of the data plot in K evaluations under the assumption that the data plot is consistent with the null hypothesis.
#' We distinguish between three different scenarios:
#' \itemize{
#' \item Scenario I: in each of K evaluations a different data set and a different set of (m-1) null plots is shown.
#' \item Scenario II: in each of K evaluations the same data set but a different set of (m-1) null plots is shown.
#' \item Scenario III: the same lineup, i.e. same data and same set of null plots, is shown to K different observers.
#' }
#' @param x number of observed picks of the data plot
#' @param K number of evaluations
#' @param m size of the lineup
#' @param N MC parameter: number of replicates on which MC probabilities are based. Higher number of replicates will decrease MC variability.
#' @param scenario numeric value, one of 1,2, or 3, indication the type of simulation used: scenario 3 assumes that the same lineup is shown in all K evaluations
#' @param xp exponent used, defaults to 1
#' @param target (vector) of integer values between 1 and m indicating the position(s) of the target plots. Only the number of targets will affect the probabilities.
#' @param upper.tail compute probabilities P(X >= x). Be aware that the use of this parameter is not consistent with the other distribution functions in base. There, a value of P(X > x) is computed for upper.tail=TRUE.
#' @return Vector/data frame. For comparison a p value based on a binomial distribution is provided as well.
#' @export
#' @examples
#' pVsim(15, 20, m=3) # triangle test
pVsim <- function(x, K, m = 20, N = 10000, scenario = 3, xp = 1, target = 1, upper.tail = TRUE) {
  deprecate_warn("1.0.0", "vinference::pVsim()", "pVis()")
  
  type <- paste("scenario", scenario, sep = "")
  freq <- get(type)(N = N, K = K, m = m, xp = xp, target = target)
  if (upper.tail) {
    sim <- sapply(x, function(y) sum(freq[as.numeric(names(freq)) >= y]))
    return(cbind(x = x, "simulated" = sim, "binom" = 1 - stats::pbinom(x - 1, size = K, prob = 1 / m)))
  } else {
    sim <- sapply(x, function(y) sum(freq[as.numeric(names(freq)) < y]))
    return(cbind(x = x, "simulated" = sim, "binom" = stats::pbinom(x - 1, size = K, prob = 1 / m)))
  }
}

#' Simulation based density for visual inference
#'
#' \lifecycle{deprecated}
#' Probablity of observing exactly x picks of the data plot in K evaluations of a lineup of size m.
#' We distinguish between three different scenarios:
#' \itemize{
#' \item Scenario I: in each of K evaluations a different data set and a different set of (m-1) null plots is shown.
#' \item Scenario II: in each of K evaluations the same data set but a different set of (m-1) null plots is shown.
#' \item Scenario III: the same lineup, i.e. same data and same set of null plots, is shown to K different observers.
#' }
#' @param x number of observed picks of the data plot
#' @param K number of evaluations of the same lineup
#' @param m size of the lineup
#' @param N MC parameter: number of replicates on which MC probabilities are based. Higher number of replicates will decrease MC variability.
#' @param scenario numeric value, one of 1, 2, or 3, indicating the type of simulation used: scenario 3 assumes that the same lineup is shown in all K evaluations
#' @param xp exponent used, defaults to 1
#' @param target location of target plot(s). By default 1. If several targets are present, specify vector of target locations.
#' @return simulation based density to observe x picks of the data plot in K evaluation under the assumption that the data plot is consistent with the null hypothesis. For comparison a p value based on a binomial distribution is provided as well.
#' @export
#' @examples
#' \dontrun{
#' dVsim(2, 20, m=3) # triangle test
#'
#' ## points in red are binomial distribution, black points are for inference
#' ## in lineups using scenario 3
#' require(ggplot2)
#' qplot(x=x, y=scenario3, data=dVsim(0:6,6,m=2)) +
#'    geom_point(aes(x,y=binom), colour="red") + ylim(c(0,0.5))
#' qplot(x=x, y=scenario3, data=dVsim(0:6,6,m=3)) +
#'    geom_point(aes(x,y=binom), colour="red") + ylim(c(0,0.5))
#' }
#'
#' # lineup with two targets: what are the probabilities to identify at least
#' # one of the targets?
#' dVsim(0:5, K=5, m=20, N=10000, scenario=3, target=1:2)
#' # slight difference between this distribution and the distribution for a
#' # lineup of size 10 with a single target:
#' dVsim(0:5, K=5, m=10, N=10000, scenario=3, target=1)
dVsim <- function(x, K, m = 20, N = 10000, scenario = 3, xp = 1, target = 1) {
  deprecate_warn("1.0.0", "vinference::dVsim()", "dVis()")
  argx <- x
  freqs <- data.frame(Var1 = 0:K)
  for (t in scenario) {
    t <- paste("scenario", t, sep = "")
    freq <- data.frame(get(t)(N = N, K = K, m = m, xp = xp, target = target))
    names(freq)[2] <- t
    freqs <- merge(freqs, freq, by = "Var1", all = T)
  }
  freq <- freqs
  freq[is.na(freq)] <- 0
  freq$binom <- stats::dbinom(0:K, size = K, prob = length(target) / m)
  names(freq)[1] <- "x"
  subset(freq, x %in% argx)
}

#' Theoretical distribution for lineups under different scenarios
#'
#' \lifecycle{deprecated}
#' @param x vector of the number of picks of the data plot out of K evaluations
#' @param K number of evaluations
#' @param m size of the lineup
#' @param scenario which scenario should be used? 1, 2, or 3?
#' @param type one of "mpfr" or "numeric". Should the result be in arbitrary numeric length or be a numeric? Internally the Rmpfr package is used to get maximal precision.
#' @export
#' @examples
#' pV(0:5, 5, m=3, scenario=3)
pV <- function(x, K, m, scenario, type = "numeric") {
  deprecate_warn("1.0.0", "vinference::pV()", "pVis()")
  res <- x
  if (3 %in% scenario) res <- cbind(res, scenario3 = pv3(x, K, m, type = type))
  if (1 %in% scenario) res <- cbind(res, scenario1 = 1 - stats::pbinom(x, size = K, prob = 1 / m) + stats::dbinom(x, size = K, prob = 1 / m))
  if (2 %in% scenario) res <- cbind(res, scenario2 = pv2(x, K, m))

  if (ncol(res) == 2) {
    res <- as.vector(res[, 2])
    names(res) <- x
    return(res)
  }
  names(res)[1] <- "x"
  res
}


#' Theoretical density for lineups under different scenarios
#'
#' \lifecycle{deprecated}
#' @param x vector of the number of picks of the data plot out of K evaluations
#' @param K number of evaluations
#' @param m size of the lineup
#' @param scenario which scenario should be used? 1, 2, or 3?
#' @param type one of "mpfr" or "numeric". Should the result be in arbitrary numeric length or be a numeric? Internally the Rmpfr package is used to get maximal precision.
#' @export
#' @importFrom lifecycle deprecate_warn
#' @examples
#' dV(0:5, 5, m=2, scenario=3)
#' ## compare to
#' dVsim(0:5, 5, m=2, scenario=3)
#'
#' require(ggplot2)
#' ## probabilities can be computed without numeric loss for K=50:
#' K <- 50
#' print(qplot(0:K, dV(0:K, K, m=20, scenario=3)));
#' print(sum(dV(0:K, K, m=20, scenario=3)));
dV <- function(x, K, m, scenario, type = "numeric") {
  deprecate_warn("1.0.0", "vinference::dV()", "dVis()")
  res <- x
  if (3 %in% scenario) res <- cbind(res, scenario3 = dv3(x, K, m, type = type))
  if (1 %in% scenario) res <- cbind(res, scenario1 = stats::dbinom(x, size = K, prob = 1 / m))
  if (2 %in% scenario) res <- cbind(res, scenario2 = dv2(x, K, m))

  if (ncol(res) == 2) {
    res <- as.vector(res[, 2])
    names(res) <- x
    return(res)
  }
  names(res)[1] <- "x"
  res
}


#' Theoretical quantile for lineups under different scenarios
#'
#' \lifecycle{deprecated}
#' @param q (vector) of the quantiles for picks of the data plot out of K evaluations
#' @param K number of evaluations
#' @param m size of the lineup
#' @param scenario which scenario should be used? 1, 2, or 3?
#' @param type one of "mpfr" or "numeric". Should the result be in arbitrary numeric length or be a numeric? Internally the Rmpfr package is used to get maximal precision.
#' @export
#' @examples
#' ## get critical values of visual triangle test:
#' qV(q=c(0.95, 0.99), K=c(5,10,15,20, 25, 30), m=3, scenario=1:3)
#'
#' ## get critical values of full lineup test:
#' qV(q=c(0.95, 0.99), K=c(5,10,15,20, 25, 30), m=20, scenario=3)
qV <- function(q, K, m, scenario, type = "numeric") {
  deprecate_warn("1.0.0", "vinference::qV()", "qVis()")
  res <- data.frame(expand.grid(q = q, K = K))
  if (1 %in% scenario) {
    # res$scenario1 <- unlist(plyr::alply(res, .margins=1, function(x) stats::qbinom(x$q, size=x$K, prob=1/m)))
    res$scenario1 <- purrr::pmap_dbl(res, function(q, K) stats::qbinom(q, size = K, prob = 1 / m))
  }
  if (2 %in% scenario) {
    res2 <- qv2(q, K, m)
    names(res2)[3] <- "scenario2"
    res <- merge(res, res2, by = c("q", "K"))
  }
  if (3 %in% scenario) {
    res3 <- qv3(q, K, m)
    names(res3)[3] <- "scenario3"
    res <- merge(res, res3, by = c("q", "K"))
  }

  res
}

dv3 <- function(x, K, m, type = "numeric") {
  T1 <- function(m) {
    if (m == 2) return(expression(1 / m * (1 / u^2 * (log((u + 1) / u) * u^2 + u - log(u + 1)))))
    if (m == 3) {
      return(expression(1 / m * (1 / u^3 * ((3 * u + 1) * log(u) - (u^3 - 3 * u - 1) * log((u + 1) / u) +
        (4 * u^3 - 3 * u - 1) * log((2 * u + 1) / (2 * u)) + 2 * u^2 -
        3 * u * log(2 * u) + log(u + 1) - log(2 * u)))))
    }
  }
  ci <- function(i, K, x) {
    res <- Rmpfr::mpfr(rep(0, length = length(i)), 120)
    idx <- which(i >= K - x)
    res[idx] <- ((-1)^(i[idx] - K + x) * Rmpfr::chooseMpfr(x, i[idx] - K + x))
    #    browser()
    res
  }
  Tone <- function(i, m) {
    #    load("data/Dis.RData")
    #    data(Dis)
    if (i == 0) return(1)
    u <- 1
    if (i == 1) return(eval(T1(m)))
    (-1)^(i - 1) / factorial(i - 1) * Dis[[m]][i] # eval(Dks[[i]])
  }
  Tone_saved <- function(i, m) {
    if (i == 0) return(1)
    #    data(T1m, package="vinference")
    T1m[[m]][i]
  }
  Tim <- function(i, m) {
    sapply(i, function(j) Tone_saved(j, m))
  }
  Dk <- function(express, name, k) {
    ## returns all k derivatives of express up to k
    res <- list()
    res[[1]] <- express # zeroeth derivative is expression itself
    for (j in 2:(k + 1)) res[[j]] <- stats::D(res[[j - 1]], name)
    res
  }
  hone <- function(x, K, m) {
    #    data(T1m, package="vinference")
    cis <- ci(0:K, K, x)

    #   choose(K, x)*sum(cis*unlist(Tim(0:K, m)))
    #   browser()
    xs <- T1m[[m]][1:K] * Rmpfr::chooseMpfr(K, x) * cis[-1]
    sum(xs) + cis[1]
  }

  if (is.null(T1m[[m]])) {
    #  if (!(m %in% c(2,3,20))) {
    stop("Not implemented for this lineup size, use bootstrap simulation with dVsim instead.")
  }
  if (K > length(T1m[[m]])) {
    if (m == 3) return(p3(x, K))
    msg <- sprintf("Not implemented for values of K > %d because of large memory needs. Use bootstrap simulation with dVsim instead.", length(T1m[[m]]))
    stop(msg)
  }
  #  Dks <- Dk(T1(m), "u", k=K)
  xs <- lapply(x, hone, K = K, m = m)
  ys <- xs[[1]]
  if (length(xs) > 1) {
    for (i in 2:length(xs)) ys <- c(ys, xs[[i]])
  }
  if (type == "mpfr") {
    return(ys)
  } else {
    return(as.numeric(ys))
  }
}

qv3 <- function(q, K, m) {
  dframe <- data.frame(expand.grid(q, K))
  names(dframe) <- c("q", "K")
  #  require(plyr)
  # res <- plyr::ddply(dframe, c("q", "K"), function(x) {
  #   hs <- cumsum(dv3(x=0:x$K, K=x$K, m=m))
  #   which(hs>=x$q)[1]
  # })
  res <- purrr::pmap_df(dframe, function(q, K) {
    hs <- cumsum(dv3(x = 0:K, K = K, m = m))
    data.frame(q = q, K = K, x = which(hs >= q)[1])
  })
  names(res)[3] <- "x"
  res$x[res$x > res$K] <- NA
  res
}


dv2 <- function(x, K, m = 3) {
  dv2one <- function(x, K, m = m) {
    g <- function(q) {
      res <- q * ((q + 2) * log(q + 2) - 2 * (q + 1) * log(q + 1) + q * log(q))
      res[q == 0] <- 0
      res
    }
    if (!(m %in% c(3, 20))) stop("Function not implemented for lineup sizes <> 3. Use bootstrap based density estimation in dVsim instead.")

    f <- function(q, K, x) choose(K, x) * g(q)^x * (1 - g(q))^(K - x)
    stats::integrate(f, 0, 1, K = K, x = x)$value
  }
  unlist(sapply(x, dv2one, K = K, m = m))
}

pv2 <- function(x, K, m = 3) {
  hdone <- function(x1, K, m) {
    sum(dv2(x1:K, K, m))
  }
  sapply(x, hdone, K = K, m = m)
}

qv2 <- function(q, K, m = 3) {
  dframe <- data.frame(expand.grid(q, K))
  names(dframe) <- c("q", "K")
  #  require(plyr)
  # res <- plyr::ddply(dframe, c("q", "K"), function(x) {
  #   hs <- cumsum(dv2(x=0:x$K, K=x$K, m=m))
  #   which(hs>=x$q)[1]
  # })
  res <- purrr::pmap_df(dframe, function(q, K) {
    hs <- cumsum(dv2(x = 0:K, K = K, m = m))
    data.frame(q = q, K = K, x = which(hs >= q)[1])
  })
  names(res)[3] <- "x"
  res$x[res$x > res$K] <- NA
  res
}



pv3 <- function(x, K, m, type = "numeric") {
  hdone <- function(x1, K, m) {
    sum(dv3(x1:K, K, m))
  }
  sapply(x, hdone, K = K, m = m)
}
